src/libcore/iter/range.rs

```diff
@@ -382,6 +382,29 @@ impl<A: Step> Iterator for ops::RangeInclusive<A> {
         None
     }
 
+    #[inline]
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    {
+        self.compute_is_empty();
+
+        let mut accum = init;
+
+        while self.start < self.end {
+            let n = self.start.add_one();
+            let n = mem::replace(&mut self.start, n);
+            accum = f(accum, n)?;
+        }
+
+        if self.start == self.end {
+            accum = f(accum, self.start.clone())?;
```

Comment:
There are [some](https://github.com/rust-lang/rust/blob/128a1fa4e1f85e04f522653bb9bee83ed6523440/src/libcore/tests/iter.rs#L1699-L1716) basic tests, but I think this could use some extra short-circuiting ones for the tricky branches in here.  (That's what went wrong in https://github.com/rust-lang/rust/pull/57974, for example.)  I'm putting the note here because it looks like this might re-iterate the end if `f` returns `Err` here.

Reply:
Good point. I had seen the tests, but had not noted this specific deviation.

It seems more consistent with the rest of the code that the value NOT be repeated if `f` returns `Err` or panics, and therefore to set `is_empty` to `Some(true)` prior to calling `f`.

There is also another bug, actually. Even if `is_empty` is `Some(true)`, `f` would still be called as long as `self.start == self.end`. I can add short-circuiting at the top, right after computing `is_empty`.

So, it seems I need to add two more tests for `try_fold` and `try_rfold`:

 - A basic test that `f` is not called after a full run: I can probably extend the existing one.
 - A different test to verify that if `f` returns `Err`, then the value is not "replayed" with the next call, and test this for a variety of values (1st, 2nd, mid, next-to-last, last).

Is there any other test you can think of?
Do you have any other test in mind?

Reply:
@scottmcm : I just pushed a change which to handle exhaustion properly. I tried to check the panic safety, but could not pass a `&mut RangeInclusive<i8>` in the closure of `catch_unwind` because the type is not `UnwindSafe`. Due to the structure of the code, the effect of a panic should be similar to that of an early return; however I am not sure how to test (or whether) to test it.

Reply:
Given the comments on #58170 which point to an agreement that a panicking callback during internal iteration need only leave the `Iterator` in a safe state, and seeing as `RangeInclusive` is necessary safe, I'll go ahead and mark this conversation as resolved.

Let me know if you think a specific test is required for this.